runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/sideinput/SideInputStateFetcherFactory.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.streaming.sideinput;
+
+import java.util.function.Function;
+import org.apache.beam.runners.dataflow.options.DataflowStreamingPipelineOptions;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill.GlobalData;
+import org.apache.beam.runners.dataflow.worker.windmill.Windmill.GlobalDataRequest;
+import org.apache.beam.sdk.annotations.Internal;
+
+/**
+ * Factory class for generating {@link SideInputStateFetcher} instances that share a {@link
+ * SideInputCache}.
+ */
+@Internal
+public final class SideInputStateFetcherFactory {
```

Comment:
i think what this allows is different implementations of fetchSideInput to be injected via work.  right now we only use fetchSideInput from the dispatcher, but if that is ever changed fetchSideInput would be different depending on implementation.

If we set both fetchSideInput and the cache when we create the SideInputStateFetcher i thought it would be a little confusing since we would either
- use the fetchSideInput the was set when SideInputStateFetcher is created and not be able to inject different fetchSideInput lambdas for different Work instances (what the existing code is before this change)
- or we could still inject fetchSideInput for different Work instances, but the initial fetchSideInput is never used
what do you think?

Reply:
No we want to use direct path for side inputs too, I didn't think of that. Let's leave as is then